Update events_cache.go

HeGaoYuan opened the pull request:
With more accurate comment

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind documentation

#### What this PR does / why we need it:
I think the origin comment is not consistent with the code.
The comment says during 10 minutes, if there are more than 10 events, then we aggregate the event.
If the event arises every 9 minutes, so during 10 minutes there are only 1 event, this is obviously not conform to the comment, so we should not aggregate the event, but the code logic will aggregate the event

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->


#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

heqg commented on staging/src/k8s.io/client-go/tools/record/events_cache.go:
```diff
@@ -37,7 +37,8 @@ const (
 	// if we see the same event that varies only by message
-	// more than 10 times in a 10 minute period, aggregate the event
+	// more than 10 times, which the max interval of two events are 10 minutes,
+	// then we aggregate the event
```
/lgtm

heqg commented:
/lgtm

heqg commented:
/ok-to-test

coffeepac commented:
/approve

HeGaoYuan commented:
/assign @serathius

leilajal commented:
/assign @caesarxuchao 
/triage accepted

caesarxuchao commented on staging/src/k8s.io/client-go/tools/record/events_cache.go:
```diff
@@ -37,7 +37,8 @@ const (
 
 	// if we see the same event that varies only by message
-	// more than 10 times in a 10 minute period, aggregate the event
+	// more than 10 times, which the max interval of two events are 10 minutes,
```
Do you mean "...more than 10 times, and the maximum interval between any two adjacent events is below 10 minutes,..."?

Then this [comment](https://github.com/kubernetes/kubernetes/blob/1945829906546caf867992669a0bfa588edf8be6/staging/src/k8s.io/client-go/tools/record/events_cache.go#L430-L434) is also inaccurate. It mentions the 10 minute "rolling interval".

HeGaoYuan replied:
Yes, that's what I mean.
I also fixed the another comment you mentioned.

HeGaoYuan committed:
Update the comment of events_cache.go

HeGaoYuan commented:
@caesarxuchao
Please have a review again.

HeGaoYuan commented:
/retest

caesarxuchao commented:
Can you add a test case to verify the behavior? e.g., verify that 10 events that are `defaultAggregateIntervalInSeconds-1` seconds apart are aggregated.

I'm considering whether it's worth fixing the code to match the original description of the comment. Perhaps that's unnecessary, as the event cache has behaved like this for years and no one has complained.

HeGaoYuan committed:
update events_cache_test.go: fix the clock bug and add more cases

HeGaoYuan commented:
> Can you add a test case to verify the behavior? e.g., verify that 10 events that are `defaultAggregateIntervalInSeconds-1` seconds apart are aggregated.
> 
> I'm considering whether it's worth fixing the code to match the original description of the comment. Perhaps that's unnecessary, as the event cache has behaved like this for years and no one has complained.

@caesarxuchao 
I've added some test case.
And I found the test code have some problems about the clock. If we use the SimpleIntervalClock and pass it to NewEventCorrelator function, each time the inside of correlator call the clock.Now() will steps the clock forward the specified duration, it is not expected. I changed the SimpleIntervalClock to FakePassiveClock, and manually control the time in the for loop.

HeGaoYuan commented:
@caesarxuchao 
I'm worried that you forgot so I remind you

serathius commented:
cc @olagacek

caesarxuchao commented on staging/src/k8s.io/client-go/tools/record/events_cache_test.go:
```diff
@@ -228,14 +240,26 @@ func TestEventCorrelator(t *testing.T) {
 			newEvent:        similarEvent,
 			expectedEvent:   setCount(similarEvent, 1),
-			intervalSeconds: defaultAggregateIntervalInSeconds,
+			intervalSeconds: defaultAggregateIntervalInSeconds + 1,
```
Why do we change this line?

caesarxuchao commented on staging/src/k8s.io/client-go/tools/record/events_cache_test.go:
```diff
@@ -228,14 +240,26 @@ func TestEventCorrelator(t *testing.T) {
+			previousEvents:  makeSimilarEvents(defaultAggregateMaxEvents, similarEvent, similarEvent.Message),
+			newEvent:        similarEvent,
+			expectedEvent:   setCount(aggregateEvent, 2),
+			intervalSeconds: defaultAggregateIntervalInSeconds - 1,
```
Can we just reuse the "similar-events-should-aggregate-event" and "similar-events-many-times-should-count-the-aggregate", just change their intervalSeconds to `defaultAggregateIntervalInSeconds-1`.

caesarxuchao commented on staging/src/k8s.io/client-go/tools/record/events_cache_test.go:
```diff
@@ -228,14 +240,26 @@ func TestEventCorrelator(t *testing.T) {
 		eventInterval := time.Duration(testInput.intervalSeconds) * time.Second
-		clock := testclocks.SimpleIntervalClock{Time: time.Now(), Duration: eventInterval}
-		correlator := NewEventCorrelator(&clock)
+		clock := testclocks.NewFakePassiveClock(time.Now())
```
Why do we change this?

caesarxuchao commented on staging/src/k8s.io/client-go/tools/record/events_cache_test.go:
```diff
@@ -338,6 +363,7 @@ func TestEventSpamFilter(t *testing.T) {
 			correlator.UpdateState(result.Event)
 		}
 
+		clock.SetTime(clock.Now().Add(eventInterval))
```
Why do we change this?
